enhancements/dns/managementstate.md

```diff
@@ -0,0 +1,126 @@
+---
+title: dns-management-state
+authors:
+  - "@rfredette"
+reviewers:
+  - "@miciah"
+  - "@danehans"
+  - "@frobware"
+  - "@sgreene570"
+  - "@knobunc"
+  - "@miheer"
+  - "@candita"
+approvers:
+  - "@miciah"
+  - "@frobware"
+  - "@danehans"
+  - "@knobunc"
+creation-date: 2021-02-09
+last-updated: 2021-02-16
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# `managementState` for DNS Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When diagnosing a DNS issue, sometimes it is helpful or even necessary to
+disable the DNS operator and patch the CoreDNS daemonset. Currently, this
+requires disabling the cluster version operator (CVO) as well so that the CVO
+doesn't re-enable the DNS operator.
+
+The DNS operator API should provide a `managementState` field, which will
+prevent the DNS operator from overwriting fields in managed deployments and
+daemonsets when the `managementState` field set to `Unmanaged` 
+
+## Motivation
+
+### Goals
+
+Provide a method to disable the DNS operator without disabling the CVO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+The DNS operator spec shall include the `managementState` field. The field will
+accept two values:
+- `Managed`: The DNS operator will manage CoreDNS configuration.
+- `Unmanaged`: The DNS operator will not manage CoreDNS configuration.
+
```

Comment:
It might be useful to have a `TimedUnmanaged`, which would keep the operator unmanaged for a specific window of time.  Then it would automatically revert at the end of the maintenance window.  The benefit would be the admin wouldn't overlook reverting the `managementState` and ending with a stale config.  The drawback is that it may not be of much value to users.

Reply:
I'm not sure how useful it would be to automatically revert to a `Managed` state, since most likely the reason that a user would set it to `Unmanaged` is that the `Managed` state is not working for whatever reason. There's probably an argument to be made that it would be useful to have as a backstop when debugging an issue, but if the operator is set to `Unmanaged` because there's a temporary patch to coredns, then it may need to stay `Unmanaged` until the fix is released, and it would be hard to pin a particular amount of time for that.

Reply:
The enhancement says that `managementState: Unmanaged` will cause the clusteroperator to report `Upgradeable=False`, and CVO will raise a warning alert if the clusteroperator continues to report `Upgradeable=False` for an hour or longer:  https://github.com/openshift/cluster-version-operator/blob/07598f405066871e3869df6a36a1eafba2598ee5/install/0000_90_cluster-version-operator_02_servicemonitor.yaml#L65-L72
That should suffice to keep the administrator from overlooking the situation.